Add Cluster Setup documentation (#4973)
openshift/odo: website/docs/getting-started/cluster-setup/kubernetes.md

---
title: Kubernetes
sidebar_position: 1
---

> # Setup a Kubernetes cluster

Reviewer: We should be consistent in the form used for the titles: **Setup a Kubernetes cluster** vs **Setting up a Kubernetes cluster**. The second form seems more appropriate to me.

*Note that this guide is only helpful in setting up a development environment; this setup is not recommended for a production environment.*

## Prerequisites

> * This guide assumes that you have a Kubernetes cluster setup, this could also be a [minikube](https://minikube.sigs.k8s.io/docs/start/) cluster.

Reviewer: Because a minikube cluster **is** a Kubernetes cluster:

```suggestion
* This guide assumes that you have a Kubernetes cluster setup, this could for example be a [minikube](https://minikube.sigs.k8s.io/docs/start/) cluster.
```

* This guide also assumes that you have admin privileges to the cluster, since the operator installation is only possible with an admin user.

> ## Enable Ingress

Reviewer: Perhaps **Enabling Ingress** ?

> To access an application externally, odo creates a URL; installing an ingress-controller helps in using this feature on a Kubernetes cluster.

Reviewer: ```suggestion
To access an application externally, you will create *URLs* using odo, which are implemented (materialized?) on a Kubernetes cluster by ingress resources; installing an ingress-controller helps in using this feature on a Kubernetes cluster.
```

To enable the Ingress feature on a Kubernetes cluster _other than minikube_, run the following command to install the default setup:
```shell

> kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v0.48.1/deploy/static/provider/cloud/deploy.yaml

Reviewer: Please indicate that this command is specific to the `nginx` ingress controller. Also, you could point to the upstream doc for more details on installation: https://kubernetes.github.io/ingress-nginx/deploy/

[//]: # (Move this section to Architecture > Service Binding or create a new Operators doc)
What are Operators?
>The Operator pattern aims to capture the key aim of a human operator who is managing a service or set of services. Human operators who look after specific applications and services have deep knowledge of how the system ought to behave, how to deploy it, and how to react if there are problems.
>
>People who run workloads on Kubernetes often like to use automation to take care of repeatable tasks. The Operator pattern captures how you can write code to automate a task beyond what Kubernetes itself provides.
> [(Source)](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/#motivation)
[//]: # (Move until here)

To install an Operator, we will first need to install OLM [(Operator Lifecycle Manager)](https://olm.operatorframework.io/) on the cluster.
```shell
curl -sL https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.18.3/install.sh | bash -s v0.18.3
```

> Running the script will take some time to install all the necessary resources in the Kubernetes cluster including the `OperatorGroup` resource,....

Reviewer: Please complete the resources list

Author: This seems like something specific to OLM, do you think it is required?

Reviewer: No, so probably just remove the 3 dots

Note: Check the OLM [release page](https://github.com/operator-framework/operator-lifecycle-manager/releases/) to use the latest version.

## Installing the Service Binding Operator
odo uses [Service Binding Operator](https://operatorhub.io/operator/service-binding-operator) to provide the `odo link` feature which helps to connect an odo component to a service or another component.

Operators can be installed in a specific namespace or across the cluster(i.e. in all the namespaces).
```shell
kubectl create -f https://operatorhub.io/install/service-binding-operator.yaml
```
Running the command will create the necessary resource in the `operators` namespace.

> If you want to access this resource from other namespaces as well, add your target namespace to `.spec.targetNamespaces` list.

Reviewer: Please specify on which file to add this

See [Verifying the Operator installation](#verifying-the-operator-installation) to ensure that the Operator was installed successfully.

> ## Installing an Operator

Reviewer: You can add a little sentence to introduce what will be done in the following steps, smth similar to:

```
To install an operator from the OperatorHub website:
```

1. Visit the [OperatorHub](https://operatorhub.io) website.
2. Search for an Operator of your choice.
3. Navigate to its detail page.
4. Click on **Install**.
5. Follow the instruction in the installation popup. Please make sure to install the Operator in your desired namespace or cluster-wide, depending on your choice and the Operator capability.

> 6. [Verifying the Operator installation](#verifying-the-operator-installation).

Reviewer: ```suggestion
6. [Verify the Operator installation](#verifying-the-operator-installation).
```

## Verifying the Operator installation
Wait for a few seconds for the Operator to install.

Once the Operator is successfully installed on the cluster, you can use `odo` to verify the Operator installation and see the CRDs associated with it; run the following command:
```shell
odo catalog list services
```
The output can look similar to:
```shell
$ odo catalog list services
Services available through Operators
NAME                                CRDs
datadog-operator.v0.6.0             DatadogAgent, DatadogMetric, DatadogMonitor
service-binding-operator.v0.9.1     ServiceBinding, ServiceBinding
```
If you do not see your installed Operator in the list, follow the [troubleshooting guide](#troubleshoot-the-operator-installation) to find the issue and debug it.

## Troubleshooting the Operator installation
There are two ways to confirm that the Operator has been installed properly.

> The examples you may see in this guide uses [Datadog Operator](https://operatorhub.io/operator/datadog-operator) and [Service Binding Operator](https://operatorhub.io/operator/service-binding-operator).

Reviewer: ```suggestion
The examples you may see in this guide use [Datadog Operator](https://operatorhub.io/operator/datadog-operator) and [Service Binding Operator](https://operatorhub.io/operator/service-binding-operator).
```

1. Verify that its pod started and is in “Running” state.
  ```shell
  kubectl get pods -n operators
  ```
The output can look similar to:
  ```shell

>   $ kubectl get csv -n operators

Reviewer: ```suggestion
  $ kubectl get pods -n operators
```

  NAME                                       READY   STATUS    RESTARTS   AGE
  datadog-operator-manager-5db67c7f4-hgb59   1/1     Running   0          2m13s
  service-binding-operator-c8d7587b8-lxztx   1/1     Running   5          6d23h
  ```

> 2. Verify that the csv is in Succeeded or Installing phase.

Reviewer: ```suggestion
2. Verify that the ClusterServiceVersion (csv) resource is in Succeeded or Installing phase.
```

---
title: Kubernetes
sidebar_position: 1
---

# Setting up a Kubernetes cluster
*Note that this guide is only helpful in setting up a development environment; this setup is not recommended for a production environment.*

## Prerequisites
* You have a Kubernetes cluster setup, this could for example be a [minikube](https://minikube.sigs.k8s.io/docs/start/) cluster.
* You have admin privileges to the cluster, since the Operator installation is only possible with an admin user.

## Enabling Ingress
To access an application externally, you will create _URLs_ using odo, which are implemented on a Kubernetes cluster by Ingress resources; installing an Ingress controller helps in using this feature on a Kubernetes cluster.

**Minikube:** To install an Ingress controller on a minikube cluster, enable the **ingress** addon with the following command:
```shell
minikube addons enable ingress
````
To learn more about ingress addon, see [the documentation on Kubernetes website](https://kubernetes.io/docs/tasks/access-application-cluster/ingress-minikube/).

> *Other Kubernetes Cluster**: To enable the Ingress feature on a Kubernetes cluster _other than minikube_, using the NGINX Ingress controller see [the official NGINX Ingress controller installation documentation](https://kubernetes.github.io/ingress-nginx/deploy/).

Reviewer: It is not clear what covers the **Other Kubernetes cluster**. It could be better to use a section:

```suggestion
### Other Kubernetes Cluster

To enable the Ingress feature on a Kubernetes cluster _other than minikube_, using the NGINX Ingress controller see [the official NGINX Ingress controller installation documentation](https://kubernetes.github.io/ingress-nginx/deploy/).
```

and idem for **minikube**

If you prefer to keep this presentation, please just add a missing \*

---
title: Kubernetes
sidebar_position: 1
---

# Setup a minikube cluster
Please note that this documentation is only useful in setting up a development environment, it is not recommended for a production environment.

This guide assumes that you have [installed minikube](https://minikube.sigs.k8s.io/docs/start/) on your system.

If you are using a Kubernetes cluster other than minikube, this guide assumes that you have admin privileges to the cluster and are logged in with the admin user; operator installation is only possible with an admin user.

**Agenda for this guide:**
* [Install the OLM](#install-the-olm)
* [Install the Service Binding Operator](#install-the-service-binding-operator)
* [Install an operator](#install-an-operator)
* [Verify the operator installation](#verify-the-operator-installation)

> 

Author: TODO: Add a note about enabling Ingress and Ingress-dns addon which is required for URL to work on minikube and perhaps kubernetes cluster.

6. [Verify the operator was installed successfully](#verify-the-operator-installation).

## Verify the Operator installation
Wait for a few seconds for the operator to install.
To verify that the operator is installed successfully and see the CRDs associated with it, run the following command.
```shell
odo catalog list services
```
The output can look similar to:
```shell
$ odo catalog list services
Services available through Operators
NAME                                CRDs
datadog-operator.v0.6.0             DatadogAgent, DatadogMetric, DatadogMonitor
service-binding-operator.v0.9.1     ServiceBinding, ServiceBinding
```
If you do not see your installed operator in the list, follow the [troubleshooting guide](#troubleshoot-the-operator-installation) to find the issue and debug it.

## Troubleshoot the Operator installation
There are two ways to confirm that the operator has been installed properly.
The examples you may see in this guide uses [Datadog Operator](https://operatorhub.io/operator/datadog-operator) and [Service Binding Operator](https://operatorhub.io/operator/service-binding-operator).
1. Verify that its pod started and is in “Running” state.  

>   ```s****************hell

Author: TODO: Fix this.

See [Verify the Operator installation](#verify-the-operator-installation) to ensure the operator is installed successfully.

## Install an Operator
1. Visit the [OperatorHub](https://operatorhub.io) website.
2. Search for an operator of your choice.
3. Navigate to its detail page.
4. Click on `Install`.
5. Follow the instruction in the installation popup.
6. [Verify the operator was installed successfully](#verify-the-operator-installation).

## Verify the Operator installation
Wait for a few seconds for the operator to install.
To verify that the operator is installed successfully and see the CRDs associated with it, run the following command.
```shell
odo catalog list services
```
The output can look similar to:
```shell
$ odo catalog list services
Services available through Operators
NAME                                CRDs
datadog-operator.v0.6.0             DatadogAgent, DatadogMetric, DatadogMonitor
service-binding-operator.v0.9.1     ServiceBinding, ServiceBinding
```
If you do not see your installed operator in the list, follow the [troubleshooting guide](#troubleshoot-the-operator-installation) to find the issue and debug it.

> ## Enable Ingress addon

Reviewer 2: I would move this section above "Install Operators". Enabling ingress is more important, as without it users won't be even able to access their applications.

---
title: Kubernetes
sidebar_position: 1
---

# Setup a minikube cluster

> Please note that this documentation is only useful in setting up a development environment, it is not recommended for a production environment.

Author: About this, there's something that came up in my discussion with Marc. I'll try to phrase it here - 
If with odo, we are trying to reduce the gap between inner loop and outer loop, which is bringing the production environment closer to the developer or vice versa, do you think it makes sense to put this warning here? Wouldn't it ward off some users who are really trying to use odo with the intention of reducing the gap?

> Please note that this documentation is only useful in setting up a development environment, it is not recommended for a production environment.

Reviewer 3: We are talking only about the k8s/ocp clusters that odo docs help spin up. We are not saying that odo is dev cluster only tool.

This guide assumes that you have [installed minikube](https://minikube.sigs.k8s.io/docs/start/) on your system.

If you are using a Kubernetes cluster other than minikube, this guide assumes that you have admin privileges to the cluster and are logged in with the admin user; operator installation is only possible with an admin user.

**Agenda for this guide:**
* [Install the OLM](#install-the-olm)
* [Install the Service Binding Operator](#install-the-service-binding-operator)
* [Install an operator](#install-an-operator)
* [Verify the operator installation](#verify-the-operator-installation)

> * [Enable Ingress addon](#enable-ingress-addon)

Reviewer 3: TBH, we don't need this Agenda because it shows up by default in the sidebar. And it's kinda sticky. Doesn't matter how much you scroll, it stays there. I would not put it separately in the doc.

## Install the OLM

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Author: Shouldn't we put this in the Operator doc?

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Reviewer 3: Yes we should. But at the moment, we are adding this doc. So we should put it here. If operator doc was ready, we could have linked to it. Once we have it ready, we could replace this paragraph with a link to it.

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Reviewer 2: Rather than coming up with our own definition of what Operator is we should link to existing Kubernetes documentation 

https://kubernetes.io/docs/concepts/extend-kubernetes/operator/

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Reviewer 3: We can and should surely link to k8s docs whenever it makes sense. But we should also make things simpler to understand for users who don't want/need to know things from k8s perspective.

Having said that, in this case, the `Motivatoin` section of the doc is pretty good and can be used to explain Operators instead of using what I suggested.

But overall, I hope that odo docs attempt to make k8s things simpler to understand for users. If that means coming up with our own definitions, so be it. Code should not be the only place where odo follows the philosophy of making k8s abstractions simple for users.

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Reviewer 2: >  Code should not be the only place where odo follows the philosophy of making k8s abstractions simple for users.

agreed 

> If that means coming up with our own definitions,

That is dangerous. We should not come up with our own definitions for stuff that we did no created. We should no try to redefine things.

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Reviewer 2: Why not use already existing definition from  https://kubernetes.io/docs/concepts/extend-kubernetes/operator/ or https://www.redhat.com/en/topics/containers/what-is-a-kubernetes-operator ? Do you think that those are too technical?

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Reviewer 3: > That is dangerous. 

The other way of looking at it is trying to do ELI5. Instead of really trying to do ELI5, we should try to explain things as if our readers have no knowledge of k8s.

> Why not use already existing definition from https://kubernetes.io/docs/concepts/extend-kubernetes/operator/ or https://www.redhat.com/en/topics/containers/what-is-a-kubernetes-operator ? Do you think that those are too technical?

I already said in my previous comment (https://github.com/openshift/odo/pull/4973#discussion_r686613855) that in this case, reusing what's in the "Motivation" section makes sense because it's written in really simple language.

> The Operator Lifecycle Manager(OLM) is a component of the Operator Framework, an open source toolkit to manage Kubernetes native applications, called Operators, in a streamlined and scalable way.[(Source)](https://olm.operatorframework.io/)

Author: I will add the excerpt of Motivation from the official k8s doc, it's short and simple as compared to the RH article.
